File: day4/solve2.cpp
#include "solve2.h"
bool process2(matrix& m, int col, int line);
int solve2(matrix& m)
{
    auto max_lines = m.size();
    auto max_col = m[0].size();

    int cc = 0;
    
    for (auto line = 0; line < max_lines; line++) {
        for (auto col = 0; col < max_col; col++) {
            cc += process2(m, col, line);
        }
    }
    
    return cc;
}

bool process2(matrix& m, int col, int line)
{
    char current = m[line][col];
    if (current != 'A') return false;
    
    // a.b
    // .X.
    // c.d
    char a = ' ', d = ' ', b = ' ', c = ' ';
    access(m, col - 1, line - 1, &a);
    access(m, col + 1, line + 1, &d);

    access(m, col + 1, line - 1, &b);
    access(m, col - 1, line + 1, &c);
    
    return (
        ((a == 'M' && d == 'S') || (a == 'S' && d == 'M'))
          &&
        ((b == 'M' && c == 'S') || (b == 'S' && c == 'M'))
    );
}

File: day06/src/main/kotlin/solve2.kt
package nodomain.aoc2024

fun solve2(matrix: List<List<String>>): Int {
    val positions = listPositions(matrix)

    var sum = 0

    val cloned = matrix.map {ArrayList(it)}
    val start = locateStarting(matrix)
    for (i in 1..<positions.size) {
        val position = positions[i]
        val row = position.first
        val col = position.second

        cloned[row][col] = "#"
        if (loop(cloned, start)) {
            sum++
        }
        cloned[row][col] = "."
    }
    return sum
}

fun listPositions(matrix: List<List<String>>): List<Pair<Int, Int>> {
    val positions = matrix.map{ line -> ArrayList(line.map { _ -> false }) }
    val out = arrayListOf<Pair<Int, Int>>()
    walk(matrix, locateStarting(matrix)) { row, col, _ ->
        if (!positions[row][col]) {
            positions[row][col] = true
            out.add(Pair(row, col))
        }
        true
    }
    return out
}

fun loop(matrix: List<List<String>>, start: Pair<Int, Int>): Boolean {
    val cc = HashMap<Int, ArrayList<Int>>()
    var res = false
    walk(matrix, start) { row, col, dir ->
        val key = row * 1000 + col
        val list = cc.getOrPut(key) { ArrayList() }
        if (list.contains(dir)) {
            res = true
            return@walk false
        } else {
            list.add(dir)
        }
        true
    }

    return res
}